paperqa/agents/search.py

```diff
@@ -0,0 +1,486 @@
+from __future__ import annotations
+
+import csv
+import hashlib
+import json
+import logging
+import os
+import pathlib
+import pickle
+import zlib
+from enum import Enum
+from io import StringIO
+from typing import Any, ClassVar
+from uuid import UUID
+
+import anyio
+from pydantic import BaseModel
+from tantivy import Document, Index, Schema, SchemaBuilder, Searcher
+from tenacity import (
+    RetryError,
+    retry,
+    retry_if_exception_type,
+    stop_after_attempt,
+    wait_random_exponential,
+)
+
+from ..docs import Docs
+from ..types import DocDetails
+from ..utils import pqa_directory
+from .models import SupportsPickle
+
+logger = logging.getLogger(__name__)
+
+PQA_INDEX_ABSOLUTE_PATHS = (
+    os.environ.get("PQA_INDEX_ABSOLUTE_PATHS", "true").lower() == "true"
+)
+
+
+class AsyncRetryError(Exception):
+    pass
+
+
+class RobustEncoder(json.JSONEncoder):
+    def default(self, obj):
+        if isinstance(obj, UUID):
+            # if the obj is uuid, we simply return the value of uuid
+            return str(obj)
+        if isinstance(obj, set):
+            return list(obj)
+        return json.JSONEncoder.default(self, obj)
+
+
+class SearchDocumentStorage(str, Enum):
+    JSON_MODEL_DUMP = "json_model_dump"
+    PICKLE_COMPRESSED = "pickle_compressed"
+    PICKLE_UNCOMPRESSED = "pickle_uncompressed"
```

Comment:
```suggestion
    JSON_MODEL_DUMP = auto()
    PICKLE_COMPRESSED = auto()
    PICKLE_UNCOMPRESSED = auto()
```

Let's use `auto` when the value doesn't matter